javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/IndyModuleTypePool.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.tooling.instrumentation.indy;
+
+import io.opentelemetry.instrumentation.api.internal.cache.Cache;
+import io.opentelemetry.javaagent.extension.instrumentation.InstrumentationModule;
+import io.opentelemetry.javaagent.tooling.muzzle.AgentTooling;
+import java.util.concurrent.ConcurrentHashMap;
+import net.bytebuddy.pool.TypePool;
+
+public class IndyModuleTypePool {
+
+  private IndyModuleTypePool() {}
+
+  /**
+   * We implement the TypePool by delegating the .class resource lookup to actual dummy {@link
+   * InstrumentationModuleClassLoader} instances. We cache those dummy instances so that repeated
+   * calls to {@link #get(ClassLoader, InstrumentationModule)} share the underlying cache for {@link
+   * net.bytebuddy.description.type.TypeDescription}s. Note that it is important that the {@link
+   * InstrumentationModuleClassLoader} only weakly references the {@link ClassLoader} being
+   * instrumented, which is also used as the key of the cache. Otherwise we would accidentally
+   * create a strong reference, preventing the instrumented {@link ClassLoader} from being ever
+   * GCed. To ensure that the cached {@link InstrumentationModuleClassLoader} is only used for
+   * resource lookups and not for classloading, it is wrapped in a {@link ResourceOnlyClassLoader}.
+   * Loading classes could otherwise cause the {@link InstrumentationModuleClassLoader} to strongly
+   * reference the instrumented {@link ClassLoader}.
+   */
+  private static final ConcurrentHashMap<
+          InstrumentationModule, Cache<ClassLoader, ResourceOnlyClassLoader>>
+      classloadersForTypePools = new ConcurrentHashMap<>();
+
+  /**
+   * Provides a {@link TypePool} which has the same lookup rules for {@link
+   * net.bytebuddy.description.type.TypeDescription}s as {@link InstrumentationModuleClassLoader}
+   * have for classes.
+   *
+   * @param instrumentedCl the classloader being instrumented (e.g. for which the {@link
+   *     InstrumentationModuleClassLoader} is being created).
+   * @param module the {@link InstrumentationModule} performing the instrumentation
+   * @return the type pool
+   */
+  public static TypePool get(ClassLoader instrumentedCl, InstrumentationModule module) {
+
+    Cache<ClassLoader, ResourceOnlyClassLoader> cacheForModule =
+        classloadersForTypePools.computeIfAbsent(module, (k) -> Cache.weak());
```

Comment:
`Cache.weak()` is a concurrent weak identity map. Using a strong reference in value to key makes key uncollectable. Could you verify that the instrumented class loader can be collected.

Reply:
I think this is ok as `InstrumentationModuleClassLoader` was also changed to use a WeakReference

Reply:
That plus preventing the cached `ClassLoader` from loading any classes was done exactly to not cause a leak here.
I know that this solution is not the cleanest, but it is the simples for now IMO.
When we at some point stabilized the classloading logic of the `InstrumentationModuleClassloader`, we can consider coming back here and replacing this "hack" with a custom `TypePool` implementation instead.

Reply:
Would it make sense to not cache the type pools and class loaders here just to keep it simple for start?

Reply:
Makes sense to me. This keeps simpler and only increases the "cost" if an `InstrumentationModule` injects lots of proxies, which doesn't really happen in reality.

I'll remove the caching and revert all other changes caused by it.